### PR TITLE
Add DeepONet implemented by Pytorch

### PR DESCRIPTION
**Pull request by Jerry-Jzy**
## what is the purpose of this change
Add DeepONet implemented by Pytorch

### CODE REVIEW

**Comment by lululxvi** [comment]
Format via black.

File: deepxde/model.py [hunk 1]
```diff
@@ -246,8 +246,17 @@ def outputs(training, inputs):
...
+
+            if isinstance(inputs,tuple):
+                multi_inputs = torch.as_tensor(inputs[0])
+                inputs_num = len(inputs)
+                for i in range(1,inputs_num):
+                    inputs_i = torch.as_tensor(inputs[i])
+                    inputs_i.requires_grad_()
+                    multi_inputs = (multi_inputs, inputs_i)
+                inputs = multi_inputs
+            #inputs = torch.as_tensor(inputs)
```
Comment (lululxvi) [inline deepxde/model.py:258]: Delete the commented lines.

File: deepxde/model.py [hunk 2]
```diff
@@ -246,8 +246,17 @@ def outputs(training, inputs):
 
         def outputs_losses(training, inputs, targets, losses_fn):
             self.net.train(mode=training)
-            inputs = torch.as_tensor(inputs)
-            inputs.requires_grad_()
+
+            if isinstance(inputs,tuple):
```
Comment (lululxvi) [inline deepxde/model.py:250]: where is `else`?

File: deepxde/model.py (hunk 1 above)
Comment (Jerry-Jzy) [reply deepxde/model.py]: Fixed

File: deepxde/model.py (hunk 2 above)
Comment (Jerry-Jzy) [reply deepxde/model.py]: Has added ``else``

File: deepxde/model.py [hunk 3]
```diff
@@ -246,8 +246,17 @@ def outputs(training, inputs):
 
         def outputs_losses(training, inputs, targets, losses_fn):
             self.net.train(mode=training)
-            inputs = torch.as_tensor(inputs)
-            inputs.requires_grad_()
+            if isinstance(inputs, tuple):
+                multi_inputs = torch.as_tensor(inputs[0])
```
Comment (lululxvi) [inline deepxde/model.py:250]: Why `inputs[0]` is special?

File: deepxde/nn/pytorch/__init__.py [hunk 4]
```diff
@@ -1,6 +1,7 @@
 """Package for pytorch NN modules."""
 
-__all__ = ["FNN", "NN", "PFNN"]
+__all__ = ["FNN", "NN", "PFNN", "DeepONetCartesianProd", "PODDeepONet"]
```
Comment (lululxvi) [inline deepxde/nn/pytorch/__init__.py:3]: Reorder the list, see TensroFlow.

File: deepxde/nn/pytorch/__init__.py [hunk 5]
```diff
@@ -1,6 +1,7 @@
...
 
 from .fnn import FNN, PFNN
 from .nn import NN
+from .deeponet import DeepONetCartesianProd, PODDeepONet
```
Comment (lululxvi) [inline deepxde/nn/pytorch/__init__.py:7]: Reorder these imports. See the example of TensorFlow

File: deepxde/nn/pytorch/deeponet.py [hunk 6]
```diff
@@ -0,0 +1,138 @@
+from .fnn import FNN
+from .nn import NN
+from .. import activations
+import torch
```
Comment (lululxvi) [inline deepxde/nn/pytorch/deeponet.py:4]:
```python
import torch

from .fnn import FNN
from .nn import NN
from .. import activations
```

File: deepxde/nn/pytorch/deeponet.py [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+            network. If ``None``, then only use POD basis as the trunk net.
+
+    References:
+        `L. Lu, X. Meng, S. Cai, Z. Mao, S. Goswami, Z. Zhang, & G. E. Karniadakis. A
+        comprehensive and fair comparison of two neural operators (with practical
+        extensions) based on FAIR data. arXiv preprint arXiv:2111.05512, 2021
+        <https://arxiv.org/abs/2111.05512>`_.
+    """
+
+    def __init__(
```
Comment (lululxvi) [inline deepxde/nn/pytorch/deeponet.py:93]: `__init__` arguments are not consistent with TensorFlow one.

File: deepxde/nn/pytorch/deeponet.py [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+        kernel_initializer,
+    ):
+        super().__init__()
+        if isinstance(activation, dict):
+            activation_branch = activation["branch"]
+            self.activation_trunk = activations.get(activation["trunk"])
+        else:
+            activation_branch = self.activation_trunk = activations.get(activation)
+
+        self.pod_basis = torch.tensor(
```
Comment (lululxvi) [inline deepxde/nn/pytorch/deeponet.py:108]: Use `torch.as_tensor` not `torch.tensor`

File: deepxde/model.py (hunk 3 above)
Comment (Jerry-Jzy) [reply deepxde/model.py:250]: Not special, just use multi_inputs = torch.as_tensor(inputs[0]) to initialize

File: deepxde/nn/pytorch/__init__.py (hunk 5 above)
Comment (Jerry-Jzy) [reply deepxde/nn/pytorch/__init__.py]: Fixed

File: deepxde/nn/pytorch/deeponet.py (hunk 6 above)
Comment (Jerry-Jzy) [reply deepxde/nn/pytorch/deeponet.py]: Fixed

File: deepxde/nn/pytorch/deeponet.py [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+        else:
+            activation_branch = self.activation_trunk = activations.get(activation)
+        if callable(layer_sizes_branch[1]):
+            # User-defined network
+            self.branch = layer_sizes_branch[1]
+        else:
+            # Fully connected network
+            self.branch = FNN(layer_sizes_branch, activation_branch, kernel_initializer)
+        self.trunk = FNN(layer_sizes_trunk, self.activation_trunk, kernel_initializer)
+        # self.b = tf.Variable(tf.zeros(1))
```
[the remark this answers is not in the document]
Comment (Jerry-Jzy) [reply deepxde/nn/pytorch/deeponet.py]: Fixed

File: deepxde/nn/pytorch/deeponet.py (hunk 7 above)
Comment (Jerry-Jzy) [reply deepxde/nn/pytorch/deeponet.py:93]: Fixed

File: deepxde/nn/pytorch/deeponet.py (hunk 8 above)
Comment (Jerry-Jzy) [reply deepxde/nn/pytorch/deeponet.py]: Fixe

File: deepxde/model.py [hunk 10]
```diff
@@ -246,8 +247,17 @@ def outputs(training, inputs):
...
+                inputs_num = len(inputs)
+                for i in range(1, inputs_num):
+                    inputs_i = torch.as_tensor(inputs[i])
+                    inputs_i.requires_grad_()
+                    multi_inputs = (multi_inputs, inputs_i)
```
Comment (lululxvi) [inline deepxde/model.py:256]: What if inputs have 3 elements? would this work?

**Comment by lululxvi** [comment]
Use a sperate PR for the learning rate decay.
